test-app/tests/integration/environment-test.js

```diff
@@ -0,0 +1,364 @@
+        deferred.reject = reject;
+      });
+
+      this.env = env;
+      this.ui = ui;
+      this.resolveFieldMeta = resolveFieldMeta;
+      this.implementationMap = {
+        validate: {
+          shorthandProperty: 'callback',
+          componentPath: class extends Component {
+            constructor(owner, args) {
+              super(owner, args);
+              setTimeout(() =>
+                this.runCallback().then(deferred.resolve, deferred.reject)
+              );
+            }
+
+            async runCallback() {
+              return this.args.config.callback.call(this.args, assert);
+            }
+          },
+        },
+        ...implementationMap,
+      };
+
+      await render(hbs`
+        <ExclaimUi
+          @ui={{this.ui}}
+          @env={{this.env}}
+          @resolveFieldMeta={{this.resolveFieldMeta}}
+          @implementationMap={{this.implementationMap}}
+        />
+      `);
+
+      await deferred.promise;
+    });
+  }
+
+  exclaimTest('simple lookups', {
+    env: { foo: 'bar' },
+    ui: {
+      async $validate(assert) {
+        assert.strictEqual(get(this.env, 'foo'), 'bar');
+
+        set(this.env, 'foo', 'baz');
+        assert.strictEqual(get(this.env, 'foo'), 'baz');
+      },
+    },
+  });
+
+  exclaimTest('simple array lookups', {
+    env: { foo: ['bar', 'baz'] },
+    ui: {
+      async $validate(assert) {
+        assert.strictEqual(get(this.env, 'foo.0'), 'bar');
+        assert.strictEqual(get(this.env, 'foo.1'), 'baz');
+      },
+    },
+  });
+
+  exclaimTest('set unknown property', {
+    env: { foo: ['bar'] },
+    ui: {
+      async $validate(assert) {
+        set(this.env, 'baz', 'bax');
+        assert.strictEqual(get(this.env, 'baz'), 'bax');
+        assert.strictEqual(get(this.env, 'foo.length'), 1);
+        set(this.env, 'foo.3', 'qux');
+        assert.strictEqual(get(this.env, 'foo.3'), 'qux');
+        assert.strictEqual(get(this.env, 'foo.length'), 4);
+      },
+    },
+  });
+
+  exclaimTest('array mutation', {
+    env: { foo: ['bar', 'baz'] },
+    ui: {
+      bound: { $bind: 'foo' },
+      async $validate(assert) {
+        set(this.env, 'foo.1', 'oops');
+        assert.strictEqual(this.config.bound[1], 'oops');
+        assert.strictEqual(get(this.env, 'foo.1'), 'oops');
+        set(this.config.bound, '0', 'oops again');
+        assert.strictEqual(get(this.env, 'foo.0'), 'oops again');
+      },
+    },
+  });
+
+  exclaimTest('HTML-safe strings', {
+    env: { foo: htmlSafe('hello') },
+    ui: {
+      async $validate(assert) {
+        assert.deepEqual(get(this.env, 'foo'), htmlSafe('hello'));
+      },
+    },
+  });
+
+  exclaimTest('simple binding resolution', {
+    env: { foo: 'bar' },
+    ui: {
+      baz: { $bind: 'foo' },
+      async $validate(assert) {
+        assert.strictEqual(get(this.config, 'baz'), 'bar');
+
+        set(this.env, 'foo', 'qux');
+        assert.strictEqual(get(this.config, 'baz'), 'qux');
```

Comment:
I think `.get`s on the config object should be able to be removed now

Reply:
Yep! These tests were largely lifted whole-cloth from the other existing ones, just having the boilerplate refactored. My plan is to come through and get rid of all the `get` calls in the next PR where I clean up the `Environment` portion of things—then I won't have to comb through and think so hard about which `get`s are necessary and which aren't, I'll be able to just drop them all.